allennlp/modules/legacy_matrix_attention.py

```diff
@@ -0,0 +1,65 @@
+"""
+A ``Module`` that takes two matrices as input and returns a matrix of attentions.
+"""
+import torch
+from allennlp.modules.similarity_functions.dot_product import DotProductSimilarity
+
+from allennlp.modules.similarity_functions.similarity_function import SimilarityFunction
+
+
+from allennlp.common import Params
+from allennlp.modules.matrix_attention import MatrixAttention
+from overrides import overrides
+
+
+class LegacyMatrixAttention(MatrixAttention):
+    '''
+    This ``Module`` takes two matrices as input and returns a matrix of attentions.
+
+    We compute the similarity between each row in each matrix and return unnormalized similarity
+    scores.  Because these scores are unnormalized, we don't take a mask as input; it's up to the
+    caller to deal with masking properly when this output is used.
+
+    By default similarity is computed with a dot product, but you can alternatively use a
+    parameterized similarity function if you wish.
+
+    This is largely similar to using ``TimeDistributed(Attention)``, except the result is
+    unnormalized.  You should use this instead of ``TimeDistributed(Attention)`` if you want to
+    compute multiple normalizations of the attention matrix.
+
+    Input:
+        - matrix_1: ``(batch_size, num_rows_1, embedding_dim)``
+        - matrix_2: ``(batch_size, num_rows_2, embedding_dim)``
+
+    Output:
+        - ``(batch_size, num_rows_1, num_rows_2)``
+
+    Parameters
+    ----------
+    similarity_function: ``SimilarityFunction``, optional (default=``DotProductSimilarity``)
+        The similarity function to use when computing the attention.
+    '''
+    def __init__(self, similarity_function: SimilarityFunction = None) -> None:
+        super(LegacyMatrixAttention, self).__init__()
+
+        self._similarity_function = similarity_function or DotProductSimilarity()
+
+    @overrides
+    def forward(self, matrix_1: torch.Tensor, matrix_2: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=arguments-differ
+        tiled_matrix_1 = matrix_1.unsqueeze(2).expand(matrix_1.size()[0],
```

Comment:
This is a copy and paste from the existing implementation. 
I can't see anyway to avoid the tiling here for the element wise operations! 
If anybody has any idea how to avoid it let me know.

Reply:
In the legacy one, we're going to have to keep the tiling, because of how similarity functions are used.  The other ones can be made more efficient, but it'll be easier to work on that part after you have tests for each of the new implementations, and we can be sure that the new, more efficient implementation still passes the tests.